Bond stereo in reactions

ricrogz opened the pull request:
This implements a new mechanism for handling stereo bonds in reactions, which does not depend on the reactant's bond directions.

**Considerations**:

* Bond directions from the reactants are ignored in favor of E/Z or CIS/TRANS stereo.

* Bond directions are still considered for the products, since it is the only way for SMARTS to describe bond stereochemistry.

* Bond stereochemistry output will be produced as CIS/TRANS stereo. StereoAtoms for these will be set in accordance to the input stereochemistry specification: E/Z input will have the highest CIP ranked neighbors set as StereoAtoms, and CIS/TRANS input will keep the same StereoAtoms (in both cases the atom indexes will be updated to the ones after the reaction).

**Mechanism**:

At the time the SMARTS products templates are copied into a molecule, any present double bonds are examined. If the complete stereo group around the bond is present in the template (i.e., there is at least one bond on each side), and none of the sides has bond directions, the bond is temporarily marked as STEREOANY in the product stub.

After the Chirality of the products has been updated, the double bonds in the product molecules get stereo bonds assigned following these rules (rules are applied from top to bottom; the first that matches decides the stereochemistry of the bond):

1. Bonds from the template that had been marked as STEREOANY are checked to see if they actually can sustain stereo, and keep STEREOANY, or else they are reset to STEREONONE.
2. If a product bond has one bond direction on each side, the neighboring atoms at the end of those bonds are set as StereoAtoms, and the stereochemistry is resolved from the directions (equal directions == TRANS, different == CIS). A single bond direction on only one side of the double bond will be ignored, and the molecule will be checked against the rules below) as there is no way to resolve the stereochemistry.
3. If none of the two prior situations happened, this means that the product template contained no usable stereochemical information for the bond, and thus, we will use the atom mapping in the reaction to try to assign the stereochemistry coming from the reactant.
4. If the bond existed in the reactant, it was also a double bond, and had a defined stereochemistry then we check if the StereoAtom neighbor pair (anchors) survived the reaction. If they didn't, we check if both non-StereoAtom neighbors (non-anchors) are defined and survived the reaction. If any of those pairs survived, we forward the stereochemistry in the reactant (both pairs are symmetric to each other, if one survived, the stereo is kept).
5. If there was a double bond in the reactant with STEREOANY, this will be forwarded to the product.
6. In any other case, the bond will be marked as STEREONONE.

The commits in the pull request are more or less complete in themselves: each commit can be built independently, but they require the previous ones to actually work as expected:

- The first commit is a small utility which makes it easier to work with E/Z bonds, in the sense that it extends the usage of `getStereoAtoms()` to get the reference atoms (the one with the highest CIP rank) of the E/Z stereo bond. This unifies the API for both E/Z and CIS/TRANS stereo bonds, since `getStereoAtoms()` will now return the right references for the stereochemistry whatever the stereo type is.
- The second commit removes copying of bond directions from the reactant to the product molecule. Also, some refactoring, mostly to make the function `setReactantBondPropertiesToProduct()` a bit more legible and avoid double searching for bond ends.
- The new stereo mechanism is implemented in the third commit.
- The updated tests in the last commit have been checked to use the same 'anchors' as before the changes, so that the checks for E/Z stereo could be translated into equivalent TRANS/CIS checks.

**Notes**:
* I thought about adding a consideration for "wildcard bonds" (~) when setting STEREOANY to double bonds in the template, but since we do not allow bond directions in reactants, the wildcard bonds would never inherit directions from them.
* In case a reaction would leave one side of the double bond with two identical branches, it would still get assigned the reactant's stereochemistry -- there is no way for the mechanism to know that it has to remove the stereochemistry label without comparing the two branches (which is slow, and should be performed by the stereochemistry assignment operations). This should not be a big problem anyway.
Exactly the same is valid for the bonds we mark as STEREOANY in the template.
* It seems that SMARTS matching of stereo bonds is broken : reactants are matched independently of bond stereo  (See 3rd case in testStereoBondIsomerization() in testReactation.cpp).

Please let me know if there is something I missed, or you can think of any other interesting test cases.

d-b-w commented on Code/GraphMol/Bond.cpp:
```diff
@@ -291,6 +291,78 @@ void Bond::setStereoAtoms(unsigned int bgnIdx, unsigned int endIdx) {
+
+const INT_VECT &Bond::getStereoAtoms() const {
+  if (!dp_stereoAtoms) {
+    const_cast<Bond *>(this)->dp_stereoAtoms = findStereoAtoms(this);
```
Have you seen examples where E/Z is set but the stereoAtoms aren't? It seems like the stereoAtoms should be set whenever the E/Z is set.

ricrogz replied:
Hmmm. I think you are wrong. Have a look at `setStereoAtoms()`: they are only required for CIS/TRANS stereo, and I think most of the E/Z stereo bonds don't ever bother to set them: after all, we know the bond will be referenced to the highest CIP neighbors... question is: which atoms are those? (that's what I wanted to solve)

Setting the stereoAtoms on calculation of the E/Z was the other option, and it would make sense to have it done at the time of stereo calculation, but in case we are just reading molecules from a file or converting a SMILES string, this would slow down things. I made the implementation in the getter to only get those atoms when they are actually needed (if they ever are).

d-b-w replied:
Oh, OK. That's interesting and makes sense. I guess I still don't have enough context for intuition about how some internal stuff works.

greglandrum replied:
Yeah, this is sorely under-documented. Getting a start on fixing that is on my "soon" todo list.

d-b-w commented on Code/GraphMol/ChemReactions/ReactionRunner.cpp:
```diff
@@ -57,8 +57,72 @@ namespace ReactionRunnerUtils {
 namespace {
 const unsigned int MatchAll = UINT_MAX;
+
+class StereoBondEndCap {
```
Some documentation about what this is or what it's for would be nice.

ricrogz replied:
Oops, I wrote a small note, but it must have gotten lost in one of the changes. I'll write something again.

d-b-w commented on Code/GraphMol/ChemReactions/testReaction.cpp:
```diff
@@ -775,8 +775,8 @@ void test8Validation() {
   /* 08/08/14
-   * This test is changed due to allowing same atom mapping muliple times in the
-   * products */
+   * This test is changed due to allowing same atom mapping multiple times in
```
This comment is not meaningful to me. Not sure about others.

ricrogz replied:
I'll remove it.

d-b-w commented on Code/GraphMol/ChemReactions/testReaction.cpp:
```diff
@@ -6878,14 +6925,258 @@ void testGithub1868() {
+  rxn->initReactantMatchers();
+
+  MOL_SPTR_VECT reacts{{reactant}};
+  const auto prods = rxn->runReactants(reacts);
```
I think there's also a `rxn->runReactant(reactant)`

ricrogz replied:
It is possible, I added some tests before looking to much into this, and then never checked back.

d-b-w commented on Code/GraphMol/ChemReactions/testReaction.cpp:
```diff
@@ -6878,14 +6925,258 @@ void testGithub1868() {
+  run_gist_reaction_tests(mols);
+
+  // Now, strip bond directions and rerun the same tests to make sure tests
+  // yield the same results without bond directions in the reactants
```
nice.

ricrogz replied:
Thanks!

ricrogz commented on Code/GraphMol/Bond.cpp:
```diff
@@ -291,6 +291,75 @@ void Bond::setStereoAtoms(unsigned int bgnIdx, unsigned int endIdx) {
+  unsigned bestCipRank = 0;
+  const Atom *bestCipRankedAtom = nullptr;
+  const auto mol = atom->getOwningMol();
+  for (ROMol::ADJ_ITER_PAIR atomIter = mol.getAtomNeighbors(atom);
```
Oh, I didn't even think about that. I'll use it when I update.

ricrogz commented:
I made the changes suggested by Dan. He also suggested some examples I might try. One of them failed because the "product bond" matched the reactant bond backwards, so I added code to account for this possiblity, plus a test that exercises it. Also some slight refactoring.

ricrogz commented on Code/GraphMol/ChemReactions/ReactionRunner.cpp:
```diff
@@ -284,6 +349,23 @@ RWMOL_SPTR convertTemplateToMol(const ROMOL_SPTR prodTemplateSptr) {
     }
+
+    // Double bond stereo: if a double bond has at least one bond on each side,
+    // and none of those has a direction, then we temporarily set STEREOANY.
```
Yes, I think it would be enough. Since we only consider bond directions in the product, if there is only one bond direction in the template, then it is not possible to find another one later on.

ricrogz commented on Code/GraphMol/ChemReactions/ReactionRunner.cpp:
```diff
@@ -361,31 +443,204 @@ ReactantProductAtomMapping *getAtomMappingsReactantProduct(
+  auto match = std::find_if(pMatches.begin(), pMatches.end(), areAtomsBonded);
+
+  if (match == pMatches.end()) {
+    throw std::logic_error("match not found");
```
Sure, I'm still not used to many RDKit idioms :)

ricrogz commented on Code/GraphMol/ChemReactions/ReactionRunner.cpp:
```diff
@@ -361,31 +443,204 @@ ReactantProductAtomMapping *getAtomMappingsReactantProduct(
+      Atom *pStart = pBond->getBeginAtom();
+      Atom *pEnd = pBond->getEndAtom();
+
+      pStart->calcImplicitValence();
```
Yeah, that makes sense. I'll add it.

ricrogz commented:
Your suggestions make sense, Greg. I have included them and updated.

ricrogz commented:
Hm. I'm having some second thoughts about the STEREOANY treatment in the product templates.

They way they are handled now means that, by default, the stereo existing in the reactant will be destroyed if we include the double bond in the reaction and we don't assign it a specific stereochemistry.

Maybe we should, by default, forward the reactant stereochemistry unless we explicitly destroy the stereo in the reaction (by specifying the double bond with two neighbors on each side, with no directions at all)? That would fix issue #31.

ricrogz commented:
An example on my previous comment and the change I am thinking about:

With the current patch, behavior is like this:

With a reactant like "C/C=C/C" and a reaction like "CC=CC >> CC=CBr", we would get the product "CC=CBr": since the product template in the reaction has 1 bond on each side of the double bond, and no bond directions are specified, the stereo in the reactant is overridden with STEREOANY.

To achieve stereo in the product, we would have to explicitly specify it in the reaction, like "CC=CC >> C/C=C/Br" or "CC=CC >> C/C=C\Br".

The option I am thinking about would be the opposite:

With the same reactant reactant ("C/C=C/C") and reaction ("CC=CC >> CC=CBr"), the product would be "C/C=C/Br", because we do not explicitly override the reactant's stereochemistry, so it gets forwarded to the product.

To override the stereochemistry and make the bond STEREOANY, we would need to write a more detailed reaction that explicitly depicts all bounds around the stereo bond, like "CC=CC >> C([H])C=C([H])Br" or "[C:1]([*:2])[C:3]=[C:4]([*:5])[C:6] >> [C:1]([*:2])[C:3]=[C:4]([C:5])[Br:6]", etc, depending on the case we are working on. In these, as we do not specify directions on any of the sides of the double bond (it would be enough to not specify then on one of the sides), we explicitly destroy the product's stereochemistry, and the bond will be set to STEREOANY.

This second option would require some more work, but I think it is feasible, and maybe also the behavior that would be reasonable to expect.

Opinions welcome :)

bp-kelley commented on Code/GraphMol/ChemReactions/ReactionRunner.cpp:
```diff
@@ -361,31 +448,208 @@ ReactantProductAtomMapping *getAtomMappingsReactantProduct(
+  auto rStartAnchorMatches = react2Prod.find(start.getAnchorIdx());
+  auto rEndAnchorMatches = react2Prod.find(end.getAnchorIdx());
+
+  // If any of the anchor pair is not mapped, see if we can make the match with
```
It might help to have an example reaction smarts of these in the comments (and below)l

ricrogz replied:
I changed this to also consider "stereoatom, non-stereoatom" pair stereo forwarding, and added a long comment with examples.

bp-kelley commented on Code/GraphMol/ChemReactions/ReactionRunner.cpp:
```diff
@@ -927,6 +1184,8 @@ void addReactantAtomsAndBonds(const ChemicalReaction &rxn, RWMOL_SPTR product,
   // hasn't been disturbed
   checkAndCorrectChiralityOfProduct(chiralAtomsToCheck, product, mapping);
 
+  updateStereoBonds(product, *reactant, mapping);
```
Needs a comment I think.

ricrogz replied:
Added.

bp-kelley commented on Code/GraphMol/ChemReactions/testReaction.cpp:
```diff
@@ -1304,18 +1301,19 @@ void test12DoubleBondStereochem() {
-  // stereochem querying stuff doesn't match C/C=C\Cl when C\C=C/Cl is provided
-  // as
-  // a query.
+  // that have stereochem indicated, is currently not completely correct since
```
Is there any real reason not to just turn on chirality for the substructure matching?

I really thought we had a while ago, but I think I had been doing this on the outside.  EnumerateLibrary does use subtract matching, but I think it doesn't really reject anything which is a bug there I think.

ricrogz replied:
I think turning off the chirality in matching is useful, at least, for debugging.

greglandrum replied:
How about adding a `SubstructMatchParameters` data member to the reaction class, then the behavior can be tuned?
https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/Substruct/SubstructMatch.h#L28

ricrogz replied:
Hum. Would all the parameters in `SubstructMatchParameters` make sense in a reaction? If not at least most of them, I would stay with the boolean flag.

bp-kelley replied:
I can see some other things we could add such as sanitize products and such which could be really convenient.

greglandrum replied:
At the moment the only two that really apply are `useChirality` and `aromaticMatchesConjugated`, but I am hoping to add a few more of the usual options (special rules for 5-membered heterocycles, etc) for tuning SSS in the near(ish) future (ideally this release cycle). If we used `SubstructMatchParameters` in the reaction code then these would automatically be used there as well. I'm not sure whether or not that's super useful, but it's probably not completely without value. :-)

bp-kelley commented on Code/GraphMol/Bond.cpp:
```diff
@@ -291,6 +291,79 @@ void Bond::setStereoAtoms(unsigned int bgnIdx, unsigned int endIdx) {
+
+const INT_VECT &Bond::getStereoAtoms() const {
+  if (!dp_stereoAtoms) {
+    const_cast<Bond *>(this)->dp_stereoAtoms = findStereoAtoms(this);
```
We should just make dp_stereoAtoms mutable at least to remove the const_cast.

This also indicates that we need to put a lock here at some point, but that is probably a bridge too far.

greglandrum replied:
I disagree strongly. 
dp_stereoAtoms shouldn't be mutable any more than the bond order should be mutable. 
This is a special case where we're initializing a value that hasn't already been initialized (because it wasn't needed), it's not making an actual substantive change to the bond itself.

There's an argument to be made that the code in Chirality.cpp which assigns STEREOE/STEREOZ should set the stereo atoms so that we don't have to do this here, but that's a different point.

greglandrum replied:
after having watched the video about the new meaning of const that Brian recommended (https://www.google.com/amp/s/herbsutter.com/2013/01/01/video-you-dont-know-const-and-mutable/amp/) and thinking about it a bit more, I'm no longer thrilled with the fact that this getter is now no longer threadsafe.

Fixing this is going to require re-thinking the way we handle `dp_stereoAtoms`. I think this means creating the `dp_stereoAtoms` structure when we set STEREOE/STEREOZ and then having a PRECONDITION() in this method that fails if `dp_stereoAtoms` is not set.

Since that's a change in behavior beyond the way stereo in reactions is handled, I would be fine if we merged this PR as is and immediately created an issue for the threadsafety change. I can then take a swing at a solution for the problem. Alternatively, if you guys think it makes more sense to get it out of the way immediately, I can do a PR against @ricrogz's branch.

ricrogz replied:
I'm ok with both solutionsI don't have any preference.

d-b-w replied:
A third option would be to limit the scope of this change to bonds that are already have `stereoAtoms`, and leave resolution of `stereoAtoms` for E/Z bonds as a separate concern. I think that would preserve the thread-safety and `const` character of ``dp_stereoAtoms`, and address most of the problems that this commit is intending to address.

Then we could decide in some other commit whether stereoAtoms is a cache (and so calculated on the fly with some appropriate locking) or an attribute (and so calculated during sanitize or some other similar rectification step).

ricrogz replied:
@d-b-w, I think limiting the scope of this case to E/Z bonds which have stereoatoms set, or checking and setting them in every place they are required would be more complicated than addressing here, even if in a temporary manner.

I'll update with a temporary fix so we can merge this being thread safe, and then we can open an issue to make a better handling of `dp_stereoAtoms`, as @greglandrum suggested.

ricrogz commented:
Ok, I found a bug a bug in my code (a typo in the `find` in the `StereoBondEndCap` constructor), and added consideration for mixed stereoatom -- non-stereoatom stereo forwarding from the reactant (I forgot about it, and only included the symmetric case), plus the comments that Brian requested.

ricrogz commented:
Added a temporary solution to make `Bond::getStereoAtoms()` threadsafe.

bp-kelley commented on Code/GraphMol/Bond.cpp:
```diff
@@ -322,48 +322,51 @@ const Atom *findHighestCIPNeighbor(const Atom *atom, const Atom *skipAtom) {
 INT_VECT &Bond::getStereoAtoms() {
   if (!dp_stereoAtoms) {
-    dp_stereoAtoms = findStereoAtoms(this);
+#ifdef RDK_THREADSAFE_SSS
```
I don't think this is the right solution.  It still has the issue of casting away const, and I feel the call_once is kind of odd here.  I think the right solution is to either set this when assigning stereochemistry, or perhaps move the calls to bond::updatePropertyCache, although I'm unsure if this is called in the correct places.

However, there may be a new bug introduced here:  I'm troubled by the fact that findStereoAtoms apparently can't be called more than once.  What if the molecule is edited?  Will findStereoAtoms every be called again?  I don't think it will, and there is no way to reset dp_stereoAtoms.

The previous code was "only" creating the INT_VECT when it didn't exist, it wasn't actually doing anything with it so I don't think we saw this issue.

bp-kelley replied:
Is there any chance we can punt this code into addReactantTemplate and addProductTemplate for now?

```
void setStereoAtoms(Bond *bond) {
  if (bond->hasOwningMol() &&
      (bond->getStereo() == Bond::BondStereo::STEREOE ||
       bond->getStereo() == Bond::BondStereo::STEREOZ)) {
    const Atom *startStereoAtom =
        findHighestCIPNeighbor(bond->getBeginAtom(), bond->getEndAtom());
    const Atom *endStereoAtom =
        findHighestCIPNeighbor(bond->getEndAtom(), bond->getBeginAtom());

     if (startStereoAtom != nullptr && endStereoAtom != nullptr) {
      int startStereoAtomIdx = static_cast<int>(startStereoAtom->getIdx());
      int endStereoAtomIdx = static_cast<int>(endStereoAtom->getIdx());

       bond->setStereoAtoms(startStereoAtomIdx, endStereoAtomIdx);
    }
  }
}

unsigned int addReactantTemplate(ROMOL_SPTR mol) {
    this->df_needsInit = true;
    for(auto bond: mol->bonds()) {
       setStereoAtoms(bond);
   }
    this->m_reactantTemplates.push_back(mol);
    return rdcast<unsigned int>(this->m_reactantTemplates.size());
  }
```

bp-kelley replied:
addReactantTemplate has the benefit of REALLY not being thread safe :)

ricrogz replied:
I have no idea how I managed, but I did read the doc for `call_once()` exactly the opposite as it is, and somehow got the idea that `once_flag` did reset after the execution.

Anyway, moving these changes to `addReactantTemplate()` seems like a good idea.

ricrogz commented:
Ok, I reverted the unfortunate attempt for thread safety, and moved `findStereoAtoms()` to ReactionRunner, and plugged it straight into the `StereoBondEndCap` class. Adding it to `addReactantTemplate()` would not work, since there it would only be applied to the bonds specified in the reaction (and those might be overridden by the product template anyway).

greglandrum commented:
> 
> When we have some more free time. I think we should assemble a bunch of reactions and make a reaction section in the docs. I'd be happy to assemble them and work on the documentation, but we are at a point where it isn't simple anymore ;)

Agreed that we should add a reaction section to the docs.
Also something on the way stereochem is handled (that one is clearly on me).

greglandrum commented:
@ricrogz I'm happy to merge this once the conflict resolved.
@d-b-w : do you agree?

ricrogz commented:
> @ricrogz I'm happy to merge this once the conflict resolved.

It should be resolved now.

d-b-w commented:
Yeah, this looks really good.

Some wonkiness with the test after the merge.

greglandrum commented:
Yeah that was a merge fail... @ricrogz ?

ricrogz commented:
> Yeah that was a merge fail... @ricrogz ?

Yes, sorry about that. My eyes were sleepy and missed that there was only one brace where there should have been two... luckily I am not a LISP developer...

It's fixed now.
